File: core/native/native_vm_engine.py
from __future__ import annotations

from typing import Any, Dict, Optional


VM_BACKENDS = ("virtualbox", "vmware", "qemu", "hyper-v")


def extract_vm_runtime(
    backend: str = "qemu",
    snapshot: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    snap = snapshot or {}
    normalized = backend.lower().replace("_", "-")

    if normalized not in VM_BACKENDS and normalized != "hyperv":
        normalized = "qemu"

    return {
        "backend": normalized,
        "vm_state": str(snap.get("vm_state", "stopped")),
        "window_streams": list(snap.get("window_streams", [])),
        "active_sessions": list(snap.get("active_sessions", [])),
        "metadata": dict(snap.get("metadata", {})),
        "bounded": True,
    }

File: src/native/nativeVmEngine.ts
/**
 * Converted from Python: core/native/native_vm_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let VM_BACKENDS: any = ["virtualbox", "vmware", "qemu", "hyper-v"];
export function extractVmRuntime(backend: any = "qemu", snapshot: any = null): any {
  var snap: any = py.or2(snapshot, () => ({}));
  var normalized: any = py.replace(String(backend).toLowerCase(), "_", "-");
  if ((!py.contains(VM_BACKENDS, normalized) && !py.eq(normalized, "hyperv"))) {
    normalized = "qemu";
  }
  return {"backend": normalized, "vm_state": py.toStr(py.get(snap, "vm_state", "stopped")), "window_streams": [...py.iter(py.get(snap, "window_streams", []))], "active_sessions": [...py.iter(py.get(snap, "active_sessions", []))], "metadata": py.pyDict(py.get(snap, "metadata", {})), "bounded": true};
}
